### PR TITLE
New state gathering

### PR DESCRIPTION
**Pull request by jhford**
Based on feedback from AWS we should be doing state updates a little differently.

### CODE REVIEW

**Comment by jhford** [comment]
So I'm seeing this working on the staging provisioner app just fine, but i'm very hesitant to promote it to production because I'm about to go on a week of PTO.  @jonasfj would you be OK handling a rollback if needed?

**Comment by jonasfj** [comment]
Sorry, I just saw this hidden in emails from github about rfcs... :) I'll take a look tomorrow... Enjoy you PTO..

File: package.json [hunk 1]
```diff
@@ -21,46 +21,46 @@
...
   },
   "dependencies": {
-    "assume": "^1.2.5",
-    "aws-sdk": "^2.23.0",
-    "azure-blob-storage": "^1.0.2",
-    "azure-entities": "^1.2.1",
-    "azure-storage": "^1.0.1",
+    "assume": "^1.5.0",
+    "aws-sdk": "^2.41.0",
+    "azure-blob-storage-temp-fork": "^1.0.4",
```
Comment (jonasfj) [inline package.json]: Not sure what you're doing here... reference a github branch is a nicer hack...

Comment (jonasfj) [inline package.json]: Not that there is any reason to change it now... Just a general recommendation... Pollution of npm is not a nice thing to do...

File: src/aws-manager.js [hunk 2]
```diff
@@ -24,6 +24,102 @@ function dateForInflux(thingy) {
...
+ * This method will run an EC2 operation.  Because the AWS-SDK client is so
+ * much fun to work with, we need to do the following things above what it does
+ * to get useful information out of it.
+ *
+ *   1. We want to have exceptions that always have region, method and service
+ *      name if available
+ *   2. Any requests which would have a requestId should include it in their
+ *      exceptions
+ *   3. Sometimes promises from AWS-SDK just magically never return and never
+ *      timeout even though we've set those options.  We have our own timeout
```
Comment (jonasfj) [inline src/aws-manager.js]: are we using the latest verison?

File: src/aws-manager.js [hunk 3]
```diff
@@ -24,6 +24,102 @@ function dateForInflux(thingy) {
...
+    // set it to the values here.
+    if (!logObj.region) {
+      logObj.region = region;
+    }
+    if (!logObj.service) {
+      logObj.service = serviceId;
+    }
+
+    // We're going to add these in because they're handy to have
+    if (!err.region) {
```
Comment (jonasfj) [inline src/aws-manager.js]: Wouldn't `object.Assign` or `_.defaults`...

File: src/aws-manager.js [hunk 4]
```diff
@@ -24,6 +24,102 @@ function dateForInflux(thingy) {
...
+  } catch (err) {
+    let logObj = {
+      //err,
+      method,
+    };
+
+    // We want to have properties we think might be in the error and
+    // are relevant right here
+    for (let prop of ['code', 'region', 'service', 'requestId']) {
+      if (err[prop]) { logObj[prop] = err[prop]; }
```
Comment (jonasfj) [inline src/aws-manager.js]: `_.defaults(logObj, _.pick(err || {},  ['code', 'region', 'service', 'requestId']))`

File: src/aws-manager.js [hunk 5]
```diff
@@ -24,6 +24,102 @@ function dateForInflux(thingy) {
...
+      if (logObj.requestId) {
+        logObj.requestIdFromHack = request.response.requestId;
+      } else {
+        logObj.requestId = request.response.requestId;
+      }
+    }
+
+    // For the region and service, if they don't already exist, we'll
+    // set it to the values here.
+    if (!logObj.region) {
```
Comment (jonasfj) [inline src/aws-manager.js]: Wouldn't `object.Assign` or `_.defaults` would give more structure...

File: package.json (hunk 1 above)
Comment (jhford) [inline package.json]: I'll delete it when it's sorted out.

File: src/aws-manager.js (hunk 2 above)
Comment (jhford) [inline src/aws-manager.js]: yes.  The `aws-sdk` is just not that good in my opinion, this is an example.  This is why I lean more towards using `aws4`; i don't trust `aws-sdk` to always actually work.

**Comment by jhford** [comment]
Adding better error information as well as removing all concurrency from the describe* calls, and doing multiple describe* calls using filters so we're making smaller API calls.

**Pull request merged**